File: app02_csharp_wiki_chat/aspnet-backend/SecureLearning.Backend/Sessions/RoomService.cs
using System.Collections.Concurrent;
using System.Security.Cryptography;

namespace SecureLearning.Backend.Sessions;

/// <summary>Pairs anonymous sessions into 2-party rooms (US-3.1); in-memory only, see <see cref="Room"/>.</summary>
public sealed class RoomService
{
    private const string CodeAlphabet = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789";
    private const int CodeLength = 8;

    private readonly ConcurrentDictionary<string, Room> rooms = new();

    public Session CreateRoom()
    {
        var room = new Room(NewRoomCode());
        rooms[room.RoomCode] = room;
        return room.AddParticipant();
    }

    public Session Join(string roomCode)
    {
        if (!rooms.TryGetValue(roomCode, out var room))
        {
            throw new RoomNotFoundException(roomCode);
        }
        return room.AddParticipant();
    }

    public bool BelongsToRoom(string roomCode, string sessionId) =>
        rooms.TryGetValue(roomCode, out var room) && room.Contains(sessionId);

    public IReadOnlyList<Session> PeersOf(string roomCode, string sessionId)
    {
        if (!rooms.TryGetValue(roomCode, out var room))
        {
            throw new RoomNotFoundException(roomCode);
        }
        return room.PeersOf(sessionId);
    }

    private static string NewRoomCode() =>
        RandomNumberGenerator.GetString(CodeAlphabet, CodeLength);
}

File: app_java_wiki_chat/spring-backend/src/main/java/com/platform/securelearning/session/RoomService.java
package com.platform.securelearning.session;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import org.springframework.stereotype.Service;

/**
 * Pairs anonymous sessions into 2-party rooms (US-3.1). Rooms are held only in
 * memory — see {@link Room} for why nothing here is persisted.
 */
@Service
public class RoomService {

    private final Map<String, Room> rooms = new ConcurrentHashMap<>();

    public Session createRoom() {
        Room room = new Room(RoomCodeGenerator.newCode());
        rooms.put(room.roomCode(), room);
        return room.addParticipant();
    }

    public Session join(String roomCode) {
        Room room = rooms.get(roomCode);
        if (room == null) {
            throw new RoomNotFoundException(roomCode);
        }
        return room.addParticipant();
    }

    public boolean belongsToRoom(String roomCode, String sessionId) {
        Room room = rooms.get(roomCode);
        return room != null && room.contains(sessionId);
    }

    public List<Session> peersOf(String roomCode, String sessionId) {
        Room room = rooms.get(roomCode);
        if (room == null) {
            throw new RoomNotFoundException(roomCode);
        }
        return room.peersOf(sessionId);
    }
}
